add codepoint-based string functions as Data.String.CodePoints

michaelficarra opened the pull request:
Functions in this module treat strings as if they were sequences of code points. I still want to pull more out of the FFI code and need to exercise all the FFI code paths by running on older browsers. But still, it should be ready for review. I'll squash after the review is done.

michaelficarra commented:
I've now tested all of the JS code paths meant for legacy browsers and moved everything I reasonably could from JS to PureScript. This should be good to go. Please review.

chexxor commented on src/Data/String/CodePoints.purs:
```diff
@@ -0,0 +1,216 @@
+codePointAtFallback n s = Array.index (toCodePointArray s) n
+
+
+count :: (CodePoint -> Boolean) -> String -> Int
```
Why does `count` require a predicate? Is it common to consider only certain characters of a code set when counting? Or is it because code points have complications, like digraphs?

michaelficarra replied:
See how `count` behaves in `Data.String`. This is just the equivalent function. It counts the number of leading code points/units that pass the predicate.

kRITZCREEK replied:
It's just a really odd name for that function.

garyb replied:
Yeah I think there was a discussion about it somewhere else. Maybe we should keep this name for now, and open an issue to rename both instances of it the next time we make a breaking change. `countPrefix` or something perhaps.

michaelficarra commented on src/Data/String/CodePoints.purs:
```diff
@@ -0,0 +1,216 @@
+module Data.String.CodePoints
```
These functions allow you to treat PureScript/JavaScript strings as if they were sequences of Unicode code points instead of their true underlying implementation, a sequence of UTF-16 code units.

hdgarrood replied:
Could you add something along those lines as a doc-comment for the module please? Just so that people who come across it on Pursuit know what it's for and when it should be used.

michaelficarra commented on src/Data/String/CodePoints.purs:
```diff
@@ -0,0 +1,216 @@
+import Data.Newtype (class Newtype)
+import Data.String as String
+-- WARN: This list must be updated to re-export any exports added to Data.String. That makes me sad.
+import Data.String (Pattern(..), Replacement(..), charAt, charCodeAt, contains, fromCharArray, joinWith, localeCompare, null, replace, replaceAll, split, stripPrefix, stripSuffix, toChar, toCharArray, toLower, toUpper, trim) as StringReExports
```
I re-export the Data.String functions so that you can switch to code point based functions by simply changing the module name. All functions that would be implemented the same (such as `null`, `contains`, `toUpper`, `trim`, etc.) are just re-exported.

hdgarrood replied:
Can you please update the comment here to clarify this? I don't think what's currently there is quite correct.

michaelficarra commented on src/Data/String/CodePoints.purs:
```diff
@@ -0,0 +1,216 @@
+derive instance newtypeCodePoint :: Newtype CodePoint _
+
+codePointFromInt :: Int -> Maybe CodePoint
+codePointFromInt n | 0 <= n && n <= 0x10FFFF = Just (CodePoint n)
```
There are 17 planes in Unicode, each with 0xFFFF code points. That means that there are 0x110000 code points in total. I will pull this value out as a constant if it is used in more than one place.

paf31 commented:
Agreed, the API looks nice. Thanks!

I don't really feel qualified to review this fully, but I'd just like to suggest that this could be a separate library (in core or otherwise), which would avoid the new dependencies on `arrays` and `lists`.

michaelficarra commented:
Unfortunately, there's not too many people who *are* qualified to review this. A reviewer would need pretty deep JS knowledge, an understanding of UTF-16, and familiarity with PureScript.

The reason I think it should be in `purescript-strings` (and the reason I re-export the `Data.String` functions) is that code point based operations are going to be preferred in almost all uses.

Do you think it would be worthwhile to use multi-parameter functions at the FFI boundary? It will probably be a bit faster, but then it adds a `Data.Function` dependency.

hdgarrood commented on test/Test/Data/String/CodePoints.purs:
```diff
@@ -0,0 +1,185 @@
+  assert $ drop 8 str == ""
+
+  log "dropWhile"
+  assert $ dropWhile (\c -> true) str == ""
```
Possibly better to use `const true` or `\_ -> true` for when we get unused variable warnings?

hdgarrood commented on test/Test/Data/String/CodePoints.purs:
```diff
@@ -0,0 +1,185 @@
+module Test.Data.String.CodePoints (testStringCodePoints) where
+
+import Prelude (Unit, discard, negate, (==), ($), (&&), (<), (<$>))
```
Open import here too please?

hdgarrood commented on src/Data/String.purs:
```diff
@@ -36,11 +36,11 @@ module Data.String
   ) where
 
-import Prelude
+import Prelude (class Ord, class Eq, class Show, Ordering(LT, EQ, GT), zero, one, (<<<), (<>), (==), ($), (-))
```
Would you mind reverting the changes to this file?

hdgarrood commented:
Re: testing other paths: I mean like in a way where people can easily rerun the same tests, ie not requiring manually testing on different browsers.

hdgarrood commented on src/Data/String/CodePoints.js:
```diff
@@ -0,0 +1,129 @@
+  return hasFromCodePoint
+    ? function (cps) {
+      if (cps.length < 10240) {
+        return String.fromCodePoint.apply(String, cps);
```
Can you please add a comment to explain what's going on here?

hdgarrood commented on src/Data/String/CodePoints.js:
```diff
@@ -0,0 +1,129 @@
+
+exports._toCodePointArray = function (fallback) {
+  return function (unsafeCodePointAt0) {
+    if (hasArrayFrom) {
```
Why have three separate code paths here? Is Array.from likely to be faster than using an iterator, which in turn is likely to be faster than the purescript fallback? Are there many platforms which provide String iteration but not Array.from?

michaelficarra replied:
@hdgarrood `Array.from` is likely faster except in cases where we are more likely to only look at earlier code points, since it requires a scan of the entire string. Of course, implementations may be doing really clever things that make this naive reasoning worthless without real world benchmarks.

I've removed some paths that, after review, I felt were unlikely to any better supported or faster than the alternative paths.

hdgarrood commented:
Unless it's prohibitively expensive to add these dependencies I agree with @michaelficarra, I think these functions ought to be here.

Also does `String` have any instances which treat strings as UTF16 code units rather than code points?

hdgarrood commented:
In fact, after this code has been in use for a while, I would like to consider renaming this module Data.String and renaming the current Data.String to Data.String.CodeUnits or something.

paf31 commented:
Looks like this shares many of the same dependencies anyway now, via `gen`, so I think it's probably not a big deal to add `arrays` and `lists` after all. `lists` only seems to be pulled in to support one recursion though, which might be possible with a fold over an array .. just a thought.

It might be worth considering how other backends would implement this API as well. cc @andyarvanitis @nwolverson

Yes, I think using `Data.Function` would be a good idea.

michaelficarra commented:
Thanks for the review, @hdgarrood! It was very helpful. I will let you know when I've updated the PR.

michaelficarra commented on src/Data/String/CodePoints.purs:
```diff
@@ -0,0 +1,215 @@
+import Data.String (Pattern(..), Replacement(..), charAt, charCodeAt, contains, fromCharArray, joinWith, localeCompare, null, replace, replaceAll, split, stripPrefix, stripSuffix, toChar, toCharArray, toLower, toUpper, trim) as StringReExports
+import Data.Tuple (Tuple(Tuple))
+import Data.Unfoldable (unfoldr)
+import Prelude (class Eq, class Ord, (&&), (||), (*), (+), (-), (<$>), (<), (<=), (<<<), (/), (<>), (==), mod)
```
Done, but now we get warnings.

michaelficarra commented:
@hdgarrood 

> Also does String have any instances which treat strings as UTF16 code units rather than code points?

Probably all of them. I'll have to do a comprehensive review. Help with this would be appreciated.

> Have you thought about how we might test the other code paths when e.g. Symbol or Array.from is not available?

Yes, and it would likely involve some sort of preprocessing phase where we set the feature test flags at the top of the file to every possible combination. I don't think it's worth the effort. This code will likely change very infrequently, and when it does we can manually do that as I did when authoring this PR.

I've made all the changes except using the uncurried functions. I'll do that tomorrow.

michaelficarra commented:
I've decided not to use the `Data.Function.Uncurried` functions for my FFI. I think this is now ready for a (final?) review.

hdgarrood commented on src/Data/String/CodePoints.js:
```diff
@@ -0,0 +1,119 @@
+  return hasFromCodePoint
+    ? function (cps) {
+      // Function.prototype.apply will fail for very large second parameters,
+      // so we don't use it for arrays with 10KB or more entries.
```
Am I missing something, or would 10K rather than 10KB be a bit more accurate here?

michaelficarra replied:
Doesn't matter to me. Anything in this general range will work fine. I'll change it.

hdgarrood replied:
I don't mean in the sense of powers of 2 vs powers of 10, I mean that we care about the number of entries in the array, which is a natural number; 10K is a natural number, but 10KB is a quantity of data, and a quantity of data doesn't make sense in this context AFAICT.

hdgarrood replied:
Anyway the updated comment looks good :+1:

hdgarrood commented on src/Data/String/CodePoints.js:
```diff
@@ -0,0 +1,119 @@
+          return function (str) {
+            var length = str.length;
+            if (index < 0 || index >= length) return Nothing;
+            if (hasArrayFrom) {
```
Sort of carrying on from our discussion before; why do we expect `Array.from` to be faster than String iteration in general? I think people will expect it to be fast to call `codePointAt` with a small argument even on very large strings, and I'm not aware of reasons to prefer `Array.from`; can you expand a little more on this please?

michaelficarra replied:
You're right, the order could be reversed here. Maybe we should do some testing and use a heuristic? The string iterator is inherently slow and hard to optimise, so I'm sure there exists a ratio between `index` and `length` such that indexing into the result of `Array.from` ends up being faster. I just don't know what that is. It will also probably vary a good bit as engines work hard to optimise iterator usage patterns.

hdgarrood replied:
> The string iterator is inherently slow and hard to optimise

Worse than implementing code point iteration manually?

michaelficarra commented:
Ping @hdgarrood @paf31. This should be ready to go. The only remaining question was related to additional optimisation work that I'm not super thrilled to take on right now. Let's merge this.

hdgarrood commented:
I'm still not really happy about that use of `Array.from`.

Also before merging, we should look through all of the `String` instances, at least those in core and contrib, to see if any need changing. If we need to use a newtype or something to have instances which treat Strings as sequences of code points, then there's a good chance that will mean significant changes to this code. If we had already released it, this would likely mean we'd have to make a ton of totally avoidable breaking changes, so let's not rush this.

On the instances question: the ones I can think of are `Eq`, `Ord`, `Show`, `Semigroup`, `Monoid` (I would do a more comprehensive search but I'm not at my computer right now). I think all of these are fine apart from perhaps `Semigroup` and `Ord`. 

For `Semigroup`, come to think of it, this instance is only problematic in the presence of lone surrogates, right? In which case I think it's fine that strings with lone surrogates might behave in slightly unexpected ways, such as failing to satisfy `length x + length y = length (x <> y)`. I'm not sure what else we'd do, apart from perhaps replace them with U+FFFD.

For `Ord` it would be ideal if it just happened to work out the same as converting to an array of code points and using the standard ordering for that; I cannot remember off the top of my head whether this is true when strings contain astral plane characters. If not, the current `Ord String` instance is at least a total ordering, so perhaps it would be best to just leave that as is and provide a newtype or a separate function for comparisons like the above.

michaelficarra commented:
@hdgarrood I will move that use of `Array.from`. But as I said above, for best performance, we should really base its usage off of some heuristic. I'll open an issue once this is merged.

On the point of the type class instances, they should all be fine except, as you point out, `Ord`. Sorting by code unit value is certainly wrong, but simply sorting by code point value is just as wrong. If we wanted a proper `Ord` implementation, we would implement [UTS10](http://unicode.org/reports/tr10/), the Unicode Collation Algorithm. Which would be built on top of a code point array, of course.

michaelficarra commented:
@hdgarrood Done.

hdgarrood commented:
Thanks.

> But as I said above, for best performance, we should really base its usage off of some heuristic.

It's not really clear to me why this might be the case.

Just a note which might come in useful later: the normal comparison operators in JS (`<`, `>` etc.), and therefore also the current `Ord String` instance, use lexicographical ordering treating strings as sequences of code units; see <https://tc39.github.io/ecma262/#sec-abstract-relational-comparison>.

In many scenarios, e.g. keys in a `Map`, it doesn't matter what the ordering given by an `Ord` instance actually is, as long as it is a total ordering. Additionally, the current instance is likely to be significantly faster than any implementation of UTS10, and also UTS10 doesn't appear to be one single algorithm but a whole family of algorithms where you have a lot of options to choose from. It's also not really clear to me whether UTS10 does actually give you a total ordering. I think therefore it might make the most sense to keep the `Ord String` instance as it is and provide UTS10 separately.

Speaking of which do you know if JavaScript is likely to start providing UTS10 comparisons any time soon?

michaelficarra commented:
@hdgarrood I think you missed my edit from above 😉. Sorry, I should have posted it as a separate message, but didn't want to spam. In short, I totally agree with you about the default `Ord` instance.

JavaScript already has `Intl.Collator`, but it was added fairly recently, and won't be available in all implementations.

So are we good to go here?

hdgarrood commented on src/Data/String/CodePoints.purs:
```diff
@@ -0,0 +1,225 @@
+
+derive instance eqCodePoint :: Eq CodePoint
+derive instance ordCodePoint :: Ord CodePoint
+derive instance newtypeCodePoint :: Newtype CodePoint _
```
I just realised that defining this instance means that people can create `CodePoint` values larger than `0x10FFFF` or smaller than 0. Presumably that's a problem?

hdgarrood commented on src/Data/String/CodePoints.purs:
```diff
@@ -0,0 +1,225 @@
+import Data.Newtype (class Newtype)
+import Data.String as String
+import Data.String.Unsafe as Unsafe
+-- WARN: This list must be updated to re-export any exports added to Data.String. That makes me sad.
```
The idea is that this module exports a superset of the things exported by `Data.String` overall, so that it's a drop-in replacement, correct? And the functions in this list are just the functions exported by `Data.String` which don't care whether a string is treated as a sequence of code points or code units? Because in that case the comment we have here currently doesn't seem entirely correct.

hdgarrood commented on src/Data/String/CodePoints.js:
```diff
@@ -0,0 +1,119 @@
+                if (o.done) return Nothing;
+                if (i === 0) return Just(unsafeCodePointAt0(o.value));
+              }
+            } else if (hasArrayFrom) {
```
Since, according to MDN, String iteration is not as well supported as `Array.from`, having `_codePointAt` first try String iteration, then `Array.from`, and then the PureScript fallback means that `codePointAt` might now be linear in either the index or the length of the entire string, depending on what's supported on the platform we're running on. I'm not necessarily opposed to using a heuristic to decide whether to use `Array.from`, but I don't think we should have the function's complexity depending on different things based on the platform. Can we just remove `Array.from` until the heuristic question has been better addressed please?

paf31 commented on src/Data/String/CodePoints.purs:
```diff
@@ -0,0 +1,225 @@
+toCodePointArrayFallback s = unfoldr decode (fromFoldable (Char.toCharCode <$> String.toCharArray s))
+  where
+  decode :: List Int -> Maybe (Tuple CodePoint (List Int))
+  decode (Cons h (Cons l rest)) | isLead h && isTrail l
```
This is the only use of `lists` here. I'd like to avoid the dependency if possible. It looks like this could be defined using `zipWith` on arrays.

michaelficarra commented:
Okay, I think we're ready for another review. You can review just the changes since last time here: https://github.com/michaelficarra/purescript-strings/compare/205838c%5E...codepoints

hdgarrood commented on src/Data/String/CodePoints.purs:
```diff
@@ -0,0 +1,273 @@
+
+-- | Returns a record with the first code point and the remaining code points
+-- | of the given string. Returns Nothing if the string is empty. Operates in
+-- | space and time linear to the length of the string.
```
Is this linear? I would have hoped it would be constant.

michaelficarra replied:
I was operating under the assumption that a naïve implementation of `String.drop` would copy the remainder of the string. But that's actually probably not the case. I can do tests, but I'm comfortable just asserting it will be constant.

hdgarrood replied:
Ok great, I'm comfortable with that too.

hdgarrood commented on src/Data/String/CodePoints.purs:
```diff
@@ -0,0 +1,273 @@
+import Data.String as String
+import Data.String.Unsafe as Unsafe
+-- WARN: In order for this module to be a drop-in replacement for Data.String,
+-- this list must be updated to re-export any exports added to Data.String.
```
If we add a new function to `Data.String` which operates on the level of code units and therefore needs a slightly different implementation in this module instead of just re-exporting, this comment isn't quite correct, right?

michaelficarra replied:
What would you suggest saying here? "any exports which don't operate on indices or Chars"? Either way, whenever Data.String is changed, there is an action item, which is what I'm trying to convey.

hdgarrood replied:
Right; how about something like:

> If a new function is added to Data.String, a version of that function should be exported from this module, which should be the same except that it should operate on the code point level rather than the code unit level. If the function's behaviour does not change based on whether we consider strings as sequences of code points or code units, it can simply be re-exported from Data.String.

hdgarrood commented on src/Data/String/CodePoints.purs:
```diff
@@ -0,0 +1,273 @@
+codePointToInt :: CodePoint -> Int
+codePointToInt (CodePoint n) = n
+
+unsurrogate :: Int -> Int -> CodePoint
```
Every use of this function has an `isLead cu0 && isTrail cu1` check beforehand; do you think it might make more sense to include this check inside the `unsurrogate` function and have it return a `Maybe CodePoint`, returning `Nothing` if it is given arguments which don't form a surrogate pair?

michaelficarra replied:
It's only an internal function, it only has two usages, and neither would be more convenient to write that way, so I'm going to leave it as-is for now.

hdgarrood replied:
Ok then :+1:

hdgarrood commented on src/Data/String/CodePoints.purs:
```diff
@@ -0,0 +1,273 @@
+
+-- | Returns the first code point of the string after dropping the given number
+-- | of code points from the beginning, if there is such a code point. Operates
+-- | in constant space and in time linear to `n`.
```
I think it might be unclear what `n` refers to here from the point of view of someone reading docs on Pursuit.

hdgarrood commented on src/Data/String/CodePoints.purs:
```diff
@@ -0,0 +1,273 @@
+
+
+-- | Returns the number of code points in the given string. Operates in
+-- | constant space and time linear to the length of the string.
```
Do you think it might be slightly clearer to say "Operates in constant space and *in* time linear to the length of the string"?

hdgarrood commented on src/Data/String/CodePoints.purs:
```diff
@@ -0,0 +1,273 @@
+
+-- | Returns a string containing the leading sequence of code points which all
+-- | match the given predicate from the given string. Operates in space and
+-- | time linear to the given number.
```
Could you please clarify what is meant by "the given number" here?

michaelficarra commented:
@hdgarrood Comments addressed.

michaelficarra commented:
@hdgarrood Updated.

hdgarrood commented on src/Data/String/CodePoints.purs:
```diff
@@ -263,10 +263,10 @@ foreign import _toCodePointArray
+toCodePointArrayFallback s = unfoldr unconsButWithTuple s
+
+unconsButWithTuple :: String -> Maybe (Tuple CodePoint String)
+unconsButWithTuple s' = (\{ head, tail } -> Tuple head tail) <$> uncons s'
```
Possibly better to just use `s` as the identifier here?

hdgarrood commented:
Ok, I think this looks good. @paf31, do you have any other comments?

paf31 commented:
👍 @hdgarrood Seems like you've reviewed this pretty thoroughly, so I'm happy to merge. Thank you @michaelficarra for adding plenty of tests, it makes it easier to see what's going on.

Is there anything in particular you'd like to add in the release notes for this change?

michaelficarra commented:
@paf31 You might want to mention that in the future we may switch this with the code unit based functions as the default.

hdgarrood commented:
:tada: Thanks very much!

hdgarrood commented:
Oops, I probably should have squashed. Oh well.

davidchambers commented:
> I probably should have squashed.

Michael intended to do so himself:

> I'll squash after the review is done.

michaelficarra commented:
Hahaha, oh boy. That's a lot of commits. Oh well, not like we can do anything about it now. Thanks, all, for the reviews and support. This wouldn't have been nearly as good without them.
